apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/Keys.cs

```diff
@@ -101,10 +98,39 @@ private KeySet(bool allKeys, IEnumerable<Key> keys, IEnumerable<KeyRange> ranges
 
         /// <inheritdoc/>
         public override bool Equals(object other)
-            => other is KeySet keySet && Protobuf.Equals(keySet.Protobuf);
+        {
+            if (other is KeySet keySet)
+            {
+                // Either both should be true or false.
+                if (AllKeys != keySet.AllKeys)
+                {
+                    return false;
+                }
+
+                // If one is null and other is not, they can't be equal.
+                if ((Keys != null && keySet.Keys == null) || (Keys == null && keySet.Keys != null))
+                {
+                    return false;
+                }
+
+                // If one is null and other is not, they can't be equal.
+                if ((Ranges != null && keySet.Ranges == null) || (Ranges == null && keySet.Ranges != null))
+                {
+                    return false;
+                }
+
+                // Now, we will have either both as null or both as non-null.
+                // TODO: Implement IEqualityComparer<T> for KeyRange in actual implementation.
+                return AllKeys == keySet.AllKeys &&
+                   ((Keys == null && keySet.Keys == null) || Keys.SequenceEqual(keySet.Keys)) &&
```

Comment:
Also, I think `SequenceEquals` requieres elemts to be on the same position and compares element references only. Not sure that's what we would want now.

Again, I think I would prefer not to have equality at all, unless we need it for something?

Reply:
SequenceEquals uses the specified-or-default equality comparer - but yes, it's ordered, which isn't great for a set. If we can get away with removing equality entirely, life becomes much simpler.